qcengine/programs/adcc.py

```diff
@@ -0,0 +1,159 @@
+"""
+Calls adcc
+"""
+import json
+import os
+import sys
+from pathlib import Path
+import importlib
+from typing import TYPE_CHECKING, Any, Dict, Optional, Tuple
+import numpy as np
+
+from qcelemental.models import AtomicResult, AtomicResultProperties, Provenance
+from qcelemental.util import deserialize, parse_version, safe_version, which_import
+
+from ..exceptions import InputError, RandomError, ResourceError, UnknownError
+from .model import ProgramHarness
+
+if TYPE_CHECKING:
+    from qcelemental.models import AtomicInput
+
+    from ..config import TaskConfig
+
+
+class AdccHarness(ProgramHarness):
+
+    _defaults = {
+        "name": "adcc",
+        "scratch": False,
+        "thread_safe": False,
+        "thread_parallel": True,
+        "node_parallel": False,
+        "managed_memory": True,
+    }
+    version_cache: Dict[str, str] = {}
+
+    class Config(ProgramHarness.Config):
+        pass
+
+    @staticmethod
+    def found(raise_error: bool = False) -> bool:
+        """Whether adcc harness is ready for operation.
+        Parameters
+        ----------
+        raise_error: bool
+            Passed on to control negative return between False and ModuleNotFoundError raised.
+        Returns
+        -------
+        bool
+            If adcc and psi4 are found, returns True.
+            If raise_error is False and adcc or psi4 is missing, returns False.
+            If raise_error is True and adcc or psi4 are missing, the error message is raised.
+        """
+        found_adcc = which_import(
+            "adcc",
+            return_bool=True,
+            raise_error=raise_error,
+            raise_msg="Please install via `conda install adcc -c adcc`.",
+        )
+        found_psi4 = which_import(
+            "psi4",
+            return_bool=True,
+            raise_error=raise_error,
+            raise_msg="Please install psi4 for adcc harness via `conda install psi4 -c psi4`.",
+        )
+        return found_adcc and found_psi4
+
+    def get_version(self) -> str:
+        """Return the currently used version of adcc"""
+        self.found(raise_error=True)
+
+        which_prog = which_import("adcc")
+        if which_prog not in self.version_cache:
+            import adcc
+
+            self.version_cache[which_prog] = safe_version(adcc.__version__)
+        return self.version_cache[which_prog]
+
+    def compute(self, input_model: "AtomicInput", config: "TaskConfig") -> "AtomicResult":
+        """
+        Runs adcc
+        """
+        self.found(raise_error=True)
+        import adcc
+        import psi4
+
+        mol = input_model.molecule
+        model = input_model.model
+        kws = input_model.keywords
```

Comment:
> I think the next step would be to discuss how/which input options are transported to the run_adc function, and what the preferred naming convention in input_model.keywords, @loriab.

Pretty much any "Parameters" or "Other Parameters" from your docs should be named _as is_ for `atomicinput.keywords`. Exceptions are items handled elsewhere in the schema like method (`model.method`), data_or_matrix (too involved), output (`protocols.stdout`). The idea is that for a harness, the user should only have to perform data layout translation, not keyword name translation.

Reply:
Well, in that case, I suggest to pass `**input_model.keywords` to `run_adc`, because then we don't have to take care of any safe-guarding in the harness, `adcc` will do it itself. `method` is already set by `model.method`, as you mentioned, and `data_or_matrix` will be forced to `wfn`.

Reply:
What's the convention wrt. to sanity checks and input handling? Some of the keywords are exclusive for example. We should have been pretty careful to die with a ValueError in that case, but that might need some translation.

Reply:
From my experience, `adcc` is pretty strict with the keyword sanity checks. What exactly do you mean by 'translation'?

Reply:
I could imagine (have not checked) that QCEngine wants error handling be done in a specific way (like use a special exception, set some flag etc.)

Reply:
right, if possible, it'd be good to sort adcc errors into input, computer resource, mysterious, etc. like https://github.com/MolSSI/QCEngine/blob/master/qcengine/programs/psi4.py#L249-L263

Reply:
Ok there are a few things we can do. We are pretty generic on `ValueError`s I just saw, meaning that a `ValueError` *can* signal that it's a problem with the input parameters, but it can also be something more internal (and thus a bug). I think the best way to solve this is is to patch adcc and ensure that during the parsing phase some specialised exception is thrown. I'll make a patch in adcc and then we can use that here and redirect to `InputError`. The C++ layer exceptions are mostly translated to `RuntimeError` and errors in the python layer use pretty much the standard python exceptions. I would maybe use UnknownError or RandomError for those?

Reply:
See https://github.com/adc-connect/adcc/pull/93.